kernel/hashlib.h

```diff
@@ -630,6 +615,15 @@ class dict
 		return !operator==(other);
 	}
 
+	unsigned int hash() const {
+		unsigned int h = mkhash_init;
+		for (auto &it : entries) {
+			h = mkhash(h, hash_ops<K>::hash(it.udata.first));
+			h = mkhash(h, hash_ops<T>::hash(it.udata.second));
+		}
+		return h;
+	}
```

Comment:
This says that two dict are different if they only differ in iterator order. ("different" meaning they might map to different hash values.) I don't think this is a reasonable definition. Especially considering that `==` is written in a way that says that two dicts are equivalent if they contain the same key-value pairs, regardless of iterator order.

Reply:
Good point.  c062047b02aa1d9fd001cb9a36b556358ccad450 should fix that.

Reply:
@cliffordwolf What do you think of https://github.com/YosysHQ/yosys/commit/c062047b02aa1d9fd001cb9a36b556358ccad450? 
Making a copy of keys and values can be expensive; however, not copying the values will require a lookup (perhaps a pointer to the value could help).
Alternatively, I'd be fine with breaking const-ness and sorting all the entries, since `dict` doesn't guarantee a specific ordering (like insertion), but instead guarantees:
https://github.com/YosysHQ/yosys/blob/d1c8837572fb343baccff4d766a8aa0bca26aab7/CodingReadme#L79-L80

Reply:
If breaking `const`ness is OK, then IMO would be even better to keep a `bool sorted` in `dict` that is set when sorted and unset when any write happens.  This way we can avoid a call to `std::sort` on an already-sorted dict every time we call `hash()`.

Reply:
> If breaking `const`ness is OK, then IMO would be even better to keep a `bool sorted` in `dict` that is set when sorted and unset when any insert/emplace happens. This way we can avoid a call to `std::sort` on an already-sorted dict every time we call `hash()`.

Not sure I agree with that. Keeping the `dict` sorted is expensive, and we would only need it in case we hash. As evident by the fact we only needed a hash function now, I think we can infer that this necessity is quite rare.

Reply:
I wonder, would it be possible to design a way of calculating the hash that is order-indepenedent (ie a linear combination of some function on keys/values) but still a good enough hash not to affect performance?

Reply:
@eddiehung Nonono, I am not proposing to keep it sorted.  Just to keep a flag that is true when it *is* sorted, because we might insert some stuff, at some point sort it when call `hash()`, then never insert to it again.  In that case, subsequent calls to `hash()` will needlessly call `std::sort` on an already sorted list of entries.  Of course, if we *do* insert/emplace to it after we set the sorted flag, we should unset that flag and not worry about sorting again until we have to do it anyway inside `hash()`.

Reply:
Sorry @boqwxp, I misread your answer. Let's revisit this after we've decided on the best approach to hashing.
@daveshah1 Thanks for the "order-independent hash" term; a quick smoogle led me to this:
https://stackoverflow.com/a/30735385, specifically
> The JDK itself proposes the following solution to this problem. The contract of the java.util.Set interface states:
>
>>Returns the hash code value for this set. The hash code of a set is defined to be the sum of the hash codes of the elements in the set, where the hash code of a null element is defined to be zero. This ensures that s1.equals(s2) implies that s1.hashCode()==s2.hashCode() for any two sets s1 and s2, as required by the general contract of Object.hashCode().
>
>An alternative to using the sum of the entries' hash codes would be to use, for example, the ^ (XOR) operator.
>
>The Scala language uses an ordering-invariant version of the Murmurhash algorithm (cf. the private scala.util.hashing.MurmurHash3 class) to implement the hashCode (or ##) method of its immutable sets and similar collections.

Summing sounds like a decent idea actually...

Reply:
Should I just do it?

Reply:
@eddiehung It's not immediately clear to me how to implement this for `dict` with the current hash function.  `mkhash()` takes an initial hash value along with the item to hash; in other words the hash value of each item is itself order-dependent.  Something like:
```
for(auto &it : entries) {
	h += mkhash(mkhash_init, hash_ops<K>::hash(it.udata.first));
	h += mkhash(mkhash_init, hash_ops<T>::hash(it.udata.second));
}
```
might work, but I assume that kind of usage is not intended, and I don't know if it is somehow problematic.

Reply:
`mkhash` is a function that combines two hashes together, and right now you're combining it with the constant `mkhash_init` which would work but likely won't give an extra value than just:
```
for(auto &it : entries) {
	h += hash_ops<K>::hash(it.udata.first);
	h += hash_ops<T>::hash(it.udata.second);
}
```

Reply:
Ah, OK, silly me.  I thought there was a need to use `mkhash` for some reason.

Reply:
@eddiehung Nope, didn't work.  5b0cb2b causes a test failure.  Should I just revert it, or is there something else you'd like to try?

Reply:
Actually, `mux.ys` is a very brittle test that keeps breaking (I can take significant blame for putting it in, that whole system didn't work out quite how I thought) so I wouldn't worry too much that it gets broken; these changes means that techmap iterates through the design in a slightly different order resulting in a slightly different circuit.

Out of interest, what OS do you have?

Reply:
I see.  I am compiling and testing on Debian 10-based [Whonix 15 workstation](https://www.whonix.org/wiki/Main_Page#overview).

Anyway I changed it to use XOR and now it passes the test just fine. ¯\_(ツ)_/¯